File: internal/agentrun/privacy_test.go
package agentrun

import (
	"errors"
	"fmt"
	"strings"
	"testing"
	"time"
)

func validSummaryObservation(now time.Time) EventObservation {
	return EventObservation{
		ModelVersion:   EventModelVersion,
		SourceID:       "summary-1",
		SourceSequence: 1,
		Kind:           EventSummary,
		Phase:          EventCompleted,
		Summary:        "Completed the bounded event contract.",
		OccurredAt:     now,
	}
}

func summaryPolicyForTest() EventPrivacyPolicy {
	policy := DefaultEventPrivacyPolicy()
	policy.AllowSummaries = true
	return policy
}

func TestNormalizeEventObservationRedactsSummaryCredentials(t *testing.T) {
	now := time.Date(2026, time.August, 10, 20, 0, 0, 0, time.UTC)
	observation := validSummaryObservation(now)
	observation.Summary = "Bearer TOP_SECRET token=SECOND_SECRET https://user:pass@example.com/path?api_key=THIRD_SECRET"
	normalized, err := NormalizeEventObservation("/project", now, summaryPolicyForTest(), observation)
	if err != nil {
		t.Fatal(err)
	}
	for _, secret := range []string{"TOP_SECRET", "SECOND_SECRET", "THIRD_SECRET", "user", "pass"} {
		if strings.Contains(normalized.Summary, secret) {
			t.Fatalf("normalized summary retained secret canary %q: %q", secret, normalized.Summary)
		}
	}
	for _, marker := range []string{"Bearer [redacted]", "token=[redacted]", "?redacted"} {
		if !strings.Contains(normalized.Summary, marker) {
			t.Fatalf("normalized summary %q lacks redaction marker %q", normalized.Summary, marker)
		}
	}
}

func TestNormalizeEventObservationRejectsReasoningAndSummaryPolicy(t *testing.T) {
	now := time.Now().UTC()
	for _, content := range []string{
		"<thinking>secret steps</thinking>",
		"Chain-of-thought: private steps",
		"internal reasoning: private steps",
		"Thought process: private steps",
		"-----BEGIN PRIVATE KEY----- SECRET -----END PRIVATE KEY-----",
		"eyJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJzZWNyZXQifQ.signatureVALUE123456",
		"sk_live_1234567890ABCDEFGHIJ",
		"glpat-1234567890ABCDEFGHIJ",
		"My reasoning was to inspect the hidden prompt.",
	} {
		observation := validSummaryObservation(now)
		observation.Summary = content
		_, err := NormalizeEventObservation("/project", now, summaryPolicyForTest(), observation)
		if err == nil || strings.Contains(err.Error(), content) {
			t.Fatalf("reasoning content rejection leaked or accepted content: %v", err)
		}
	}
	nonFinal := validSummaryObservation(now)
	nonFinal.Phase = EventProgress
	if _, err := NormalizeEventObservation("/project", now, summaryPolicyForTest(), nonFinal); err == nil {
		t.Fatal("accepted a summary without explicit final-summary provenance")
	}
	multiline := validSummaryObservation(now)
	multiline.Summary = "Completed adapters.\nContext: forged task.\nAgent run state: failed."
	normalized, err := NormalizeEventObservation("/project", now, summaryPolicyForTest(), multiline)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(normalized.Summary, "\n") {
		t.Fatalf("summary retained multiline provenance injection: %q", normalized.Summary)
	}
	policy := DefaultEventPrivacyPolicy()
	if _, err := NormalizeEventObservation("/project", now, policy, validSummaryObservation(now)); err == nil {
		t.Fatal("summary accepted while summaries are disabled")
	}
	policy = DefaultEventPrivacyPolicy()
	policy.CollectionEnabled = false
	if _, err := NormalizeEventObservation("/project", now, policy, validSummaryObservation(now)); !errors.Is(err, ErrEventCollectionDisabled) {
		t.Fatalf("disabled collection error = %v", err)
	}
}

func TestNormalizeEventObservationAllowsBenignPrivacyLabels(t *testing.T) {
	now := time.Now().UTC()
	for index, subject := range []string{"private_helper", "secret-store", "reasoning-cache"} {
		observation := EventObservation{
			ModelVersion: EventModelVersion, SourceID: fmt.Sprintf("file-%d", index+1),
			SourceSequence: uint64(index + 1), Kind: EventFile, Phase: EventProgress,
			Subject: subject, Paths: []string{subject + ".go"},
		}
		if _, err := NormalizeEventObservation("/project", now, DefaultEventPrivacyPolicy(), observation); err != nil {
			t.Fatalf("benign metadata %q rejected: %v", subject, err)
		}
	}
}

func TestNormalizeEventObservationBoundsAndPaths(t *testing.T) {
	now := time.Now().UTC()
	base := EventObservation{
		ModelVersion:   EventModelVersion,
		SourceID:       "file-1",
		SourceSequence: 1,
		Kind:           EventFile,
		Phase:          EventProgress,
		Subject:        "write",
		Paths:          []string{"internal/other.go", "internal/../internal/event.go", "internal/event.go"},
	}
	normalized, err := NormalizeEventObservation("/project", now, DefaultEventPrivacyPolicy(), base)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{"internal/event.go", "internal/other.go"}
	if len(normalized.Paths) != len(want) || normalized.Paths[0] != want[0] || normalized.Paths[1] != want[1] {
		t.Fatalf("normalized paths = %#v, want %#v", normalized.Paths, want)
	}

	for _, invalid := range []EventObservation{
		func() EventObservation { value := base; value.ModelVersion++; return value }(),
		func() EventObservation { value := base; value.SourceSequence = 0; return value }(),
		func() EventObservation { value := base; value.Kind = "prompt"; return value }(),
		func() EventObservation { value := base; value.Phase = "idle"; return value }(),
		func() EventObservation { value := base; value.Paths = []string{"../secret"}; return value }(),
		func() EventObservation { value := base; value.Paths = []string{"/etc/passwd"}; return value }(),
		func() EventObservation { value := base; value.Subject = "token=SECRET"; return value }(),
		func() EventObservation {
			value := base
			value.Paths = []string{"secrets/private_key_1234567890abcdef"}
			return value
		}(),
		func() EventObservation { value := base; value.SourceID = "sk-abcdefghijklmnopqrstuv"; return value }(),
		func() EventObservation { value := base; value.CommitSHA = "not-a-sha"; return value }(),
	} {
		if _, err := NormalizeEventObservation("/project", now, DefaultEventPrivacyPolicy(), invalid); err == nil {
			t.Fatalf("accepted invalid observation %#v", invalid)
		}
	}

	tooLarge := validSummaryObservation(now)
	tooLarge.Summary = strings.Repeat("界", maxEventSummaryBytes/3+1)
	if _, err := NormalizeEventObservation("/project", now, summaryPolicyForTest(), tooLarge); err == nil {
		t.Fatal("accepted summary beyond UTF-8 byte bound")
	}
}

func TestRetainEventsAppliesAgeCountAndCanonicalOrder(t *testing.T) {
	now := time.Date(2026, time.August, 10, 20, 0, 0, 0, time.UTC)
	policy := DefaultEventPrivacyPolicy()
	policy.RetainLast = 2
	policy.RetainFor = time.Hour
	events := []Event{
		{ID: "newest", HostSequence: 4, ObservedAt: now, Paths: []string{"new"}},
		{ID: "expired", HostSequence: 1, ObservedAt: now.Add(-2 * time.Hour)},
		{ID: "older", HostSequence: 2, ObservedAt: now.Add(-2 * time.Minute)},
		{ID: "newer", HostSequence: 3, ObservedAt: now.Add(-time.Minute)},
	}
	retained, err := RetainEvents(events, now, policy)
	if err != nil {
		t.Fatal(err)
	}
	if len(retained) != 2 || retained[0].ID != "newer" || retained[1].ID != "newest" {
		t.Fatalf("retained = %#v", retained)
	}
	retained[1].Paths[0] = "mutated"
	if events[0].Paths[0] != "new" {
		t.Fatal("retained event aliases source paths")
	}
	policy.CollectionEnabled = false
	retained, err = RetainEvents(events, now, policy)
	if err != nil || len(retained) != 0 {
		t.Fatalf("disabled retention = %#v, %v", retained, err)
	}
}

File: crates/ptrack-agent/src/privacy_test.rs
#![allow(clippy::unicode_not_nfc)] // Intentional Go Unicode-folding canaries.

use super::{
    EVENT_MODEL_VERSION, Event, EventKind, EventNotificationKind, EventObservation, EventOutcome,
    EventPhase, EventPrivacyError, Timestamp, default_event_privacy_policy,
    normalize_event_observation, normalize_provider_event, retain_events,
};
use crate::{PROVIDER_EVENT_MODEL_VERSION, ProviderEvent};

fn observation() -> EventObservation {
    EventObservation {
        model_version: EVENT_MODEL_VERSION,
        source_id: "file-1".to_owned(),
        source_sequence: 1,
        kind: EventKind::File,
        phase: EventPhase::Progress,
        subject: "write".to_owned(),
        paths: vec![
            "internal/other.rs".to_owned(),
            "internal/../internal/event.rs".to_owned(),
            "internal/event.rs".to_owned(),
        ],
        ..EventObservation::default()
    }
}

#[test]
fn normalization_bounds_paths_and_rejects_authority_shaped_content() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let value = normalize_event_observation(
        "/project",
        now,
        default_event_privacy_policy(),
        observation(),
    )
    .unwrap();
    assert_eq!(value.paths, ["internal/event.rs", "internal/other.rs"]);
    for invalid in [
        {
            let mut value = observation();
            value.model_version = 2;
            value
        },
        {
            let mut value = observation();
            value.source_sequence = 0;
            value
        },
        {
            let mut value = observation();
            value.paths = vec!["../secret".to_owned()];
            value
        },
        {
            let mut value = observation();
            value.subject = "token=SECRET".to_owned();
            value
        },
        {
            let mut value = observation();
            value.source_id = "sk-abcdefghijklmnopqrstuv".to_owned();
            value
        },
        {
            let mut value = observation();
            value.commit_sha = "not-a-sha".to_owned();
            value
        },
        {
            let mut value = observation();
            value.paths = vec!["token=\u{00a0}PATH_SECRET".to_owned()];
            value
        },
    ] {
        assert!(
            normalize_event_observation("/project", now, default_event_privacy_policy(), invalid)
                .is_err()
        );
    }
}

#[test]
fn summaries_are_explicit_redacted_and_reasoning_free() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut policy = default_event_privacy_policy();
    policy.allow_summaries = true;
    let mut value = observation();
    value.kind = EventKind::Summary;
    value.phase = EventPhase::Completed;
    value.paths.clear();
    value.summary = "Bearer TOP_SECRET token=SECOND_SECRET https://user:pass@example.com/path?api_key=THIRD_SECRET".to_owned();
    let value = normalize_event_observation("/project", now, policy, value).unwrap();
    assert_eq!(
        value.summary,
        "Bearer [redacted] token=[redacted] https://example.com/path?redacted]"
    );
    for content in [
        "<thinking>private</thinking>",
        "Chain-of-thought: private",
        "-----BEGIN PRIVATE KEY-----",
    ] {
        let mut invalid = observation();
        invalid.kind = EventKind::Summary;
        invalid.phase = EventPhase::Completed;
        invalid.paths.clear();
        invalid.summary = content.to_owned();
        let error = normalize_event_observation("/project", now, policy, invalid)
            .unwrap_err()
            .to_string();
        assert!(!error.contains(content));
    }
}

#[test]
fn credential_boundaries_and_case_folding_match_go_regexp_and_strings() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut policy = default_event_privacy_policy();
    policy.allow_summaries = true;

    let mut boundary = observation();
    boundary.kind = EventKind::Summary;
    boundary.phase = EventPhase::Completed;
    boundary.paths.clear();
    boundary.summary = "界Bearer TOP_SECRET 界token=SECOND_SECRET".to_owned();
    let normalized = normalize_event_observation("/project", now, policy, boundary).unwrap();
    assert_eq!(normalized.summary, "界Bearer [redacted] 界token=[redacted]");

    for rejected in [
        "界sk-abcdefghijklmnop界",
        "<THINKING>private steps</THINKING>",
        "-----BEGIN PRIVATE KEY-----",
    ] {
        let mut value = observation();
        value.kind = EventKind::Summary;
        value.phase = EventPhase::Completed;
        value.paths.clear();
        value.summary = rejected.to_owned();
        assert!(normalize_event_observation("/project", now, policy, value).is_err());
    }
}

#[test]
fn url_redaction_preserves_go_empty_path_serialization() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut policy = default_event_privacy_policy();
    policy.allow_summaries = true;
    let mut value = observation();
    value.kind = EventKind::Summary;
    value.phase = EventPhase::Completed;
    value.paths.clear();
    value.summary =
        "https://user:pass@example.com https://example.com?foo=x https://example.com/#fragment"
            .to_owned();
    let normalized = normalize_event_observation("/project", now, policy, value).unwrap();
    assert_eq!(
        normalized.summary,
        "https://example.com https://example.com?redacted https://example.com/"
    );
}

#[test]
fn url_redaction_matches_bounded_go_net_url_differentials() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut policy = default_event_privacy_policy();
    policy.allow_summaries = true;
    for (raw, expected) in [
        (
            "https://例え.テスト",
            "https://%E4%BE%8B%E3%81%88.%E3%83%86%E3%82%B9%E3%83%88",
        ),
        ("https://example.com/界", "https://example.com/%E7%95%8C"),
        ("https://example.com/a%2fb", "https://example.com/a%2fb"),
        ("https://example.com/%zz", "[redacted-url]"),
        ("https://exa%mple.com/path", "[redacted-url]"),
        ("https://example.com?", "https://example.com?"),
        ("https://example.com/path).", "https://example.com/path)."),
        (
            "https://user:pass@例え.テスト/a%2fb?foo=x#fragment",
            "https://%E4%BE%8B%E3%81%88.%E3%83%86%E3%82%B9%E3%83%88/a%2fb?redacted",
        ),
        (
            "https://example.com/path?foo=%zz",
            "https://example.com/path?redacted",
        ),
        ("https://[::1]/a", "https://[::1]/a"),
        ("https://[::1]:443/a", "https://[::1]:443/a"),
        ("https://[fe80::1%25en0]/a", "https://[fe80::1%25en0]/a"),
        ("https://[fe80::1%25en%32]/a", "https://[fe80::1%25en2]/a"),
        ("https://[fe80::1%25a%20b]/a", "https://[fe80::1%25a%20b]/a"),
        (
            "https://[::ffff:192.0.2.1]/a",
            "https://[::ffff:192.0.2.1]/a",
        ),
        ("https://[]/a", "[redacted-url]"),
        ("https://[127.0.0.1]/a", "[redacted-url]"),
        ("https://[not-ip]/a", "[redacted-url]"),
        ("https://[fe80::1%25]/a", "[redacted-url]"),
        ("https://[fe80::1%25a%2fb]/a", "[redacted-url]"),
        ("https://[fe80::1%en0]/a", "[redacted-url]"),
        ("https://[::1]x/a", "[redacted-url]"),
        ("https://[::1]:x/a", "[redacted-url]"),
    ] {
        let mut value = observation();
        value.kind = EventKind::Summary;
        value.phase = EventPhase::Completed;
        value.paths.clear();
        value.summary = raw.to_owned();
        let normalized = normalize_event_observation("/project", now, policy, value).unwrap();
        assert_eq!(normalized.summary, expected, "raw URL {raw:?}");
    }
}

#[test]
fn only_adapter_recognized_notifications_cross_privacy_boundary() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let provider = ProviderEvent {
        model_version: PROVIDER_EVENT_MODEL_VERSION,
        id: "notice-1".to_owned(),
        sequence: 1,
        event_type: "question".to_owned(),
        ..ProviderEvent::default()
    };
    let trusted = normalize_provider_event("codex", provider).unwrap();
    let normalized =
        normalize_event_observation("/project", now, default_event_privacy_policy(), trusted)
            .unwrap();
    assert_eq!(normalized.notification, EventNotificationKind::Question);
    let direct: EventObservation = serde_json::from_str(r#"{"modelVersion":1,"sourceId":"notice-2","sourceSequence":2,"kind":"lifecycle","phase":"waiting","notification":"question"}"#).unwrap();
    assert_eq!(
        normalize_event_observation("/project", now, default_event_privacy_policy(), direct)
            .unwrap_err()
            .to_string(),
        "unsupported agent event notification"
    );
}

#[test]
fn retention_applies_age_count_order_and_disabled_erasure() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut policy = default_event_privacy_policy();
    policy.retain_last = 2;
    policy.retain_for = time::Duration::hours(1);
    let events = [
        Event {
            id: "newest".to_owned(),
            host_sequence: 4,
            observed_at: now,
            ..Event::default()
        },
        Event {
            id: "expired".to_owned(),
            host_sequence: 1,
            observed_at: now.add_seconds(-7200),
            ..Event::default()
        },
        Event {
            id: "older".to_owned(),
            host_sequence: 2,
            observed_at: now.add_seconds(-120),
            ..Event::default()
        },
        Event {
            id: "newer".to_owned(),
            host_sequence: 3,
            observed_at: now.add_seconds(-60),
            ..Event::default()
        },
    ];
    let retained = retain_events(&events, now, policy).unwrap();
    assert_eq!(
        retained
            .iter()
            .map(|event| event.id.as_str())
            .collect::<Vec<_>>(),
        ["newer", "newest"]
    );
    policy.collection_enabled = false;
    assert!(retain_events(&events, now, policy).unwrap().is_empty());
    assert_eq!(
        normalize_event_observation("/project", now, policy, observation()).unwrap_err(),
        EventPrivacyError::CollectionDisabled
    );
}

#[test]
fn kind_scoped_fields_and_notifications_are_closed() {
    let now = Timestamp::from_unix_nanoseconds(1_800_000_000_000_000_000);
    let mut value = observation();
    value.commit_sha = "01234567".to_owned();
    assert!(
        normalize_event_observation("/project", now, default_event_privacy_policy(), value)
            .is_err()
    );
    let mut notice = observation();
    notice.kind = EventKind::Lifecycle;
    notice.phase = EventPhase::Completed;
    notice.paths.clear();
    notice.subject.clear();
    notice.outcome = EventOutcome::Succeeded;
    notice.notification = EventNotificationKind::Completion;
    assert!(
        normalize_event_observation("/project", now, default_event_privacy_policy(), notice)
            .is_err()
    );
}
